bundles/org.openhab.misc.addonsuggestionfinder/src/main/java/org/openhab/misc/addonsuggestionfinder/AddonSuggestionInfoProvider.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2010-2023 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.misc.addonsuggestionfinder;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.addon.AddonInfo;
+import org.openhab.core.addon.AddonInfoList;
+import org.openhab.core.addon.AddonInfoListReader;
+import org.openhab.core.addon.AddonInfoProvider;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * The {@link AddonSuggestionInfoProvider} provides a list of candidate suggested addons to be installed.
+ *
+ * @author Andrew Fiddian-Green - Initial contribution
+ */
+@NonNullByDefault
+@Component(name = "addon-suggestion-info-provider", service = AddonInfoProvider.class)
+public class AddonSuggestionInfoProvider implements AddonInfoProvider {
+
+    private final Set<AddonInfo> candidateAddonInfos = new HashSet<>();
+
+    @Activate
+    public AddonSuggestionInfoProvider() {
+        setCandidates(getResourceXml());
+    }
+
+    @Override
+    public @Nullable AddonInfo getAddonInfo(@Nullable String id, @Nullable Locale locale) {
+        return candidateAddonInfos.stream().filter(a -> a.getId().equals(id)).findFirst().orElse(null);
+    }
+
+    @Override
+    public Set<AddonInfo> getAddonInfos(@Nullable Locale locale) {
+        return candidateAddonInfos;
+    }
+
+    private String getResourceXml() {
+        ClassLoader loader = getClass().getClassLoader();
+        if (loader != null) {
+            InputStream stream = loader.getResourceAsStream("addons.xml");
+            if (stream != null) {
+                try {
+                    return new String(stream.readAllBytes(), StandardCharsets.UTF_8);
+                } catch (IOException e) {
+                }
+            }
+        }
+        throw new IllegalStateException("Error loading 'addons.xml' resource");
+    }
+
+    public void setCandidates(String xml) {
+        candidateAddonInfos.clear();
+        AddonInfoListReader reader = new AddonInfoListReader();
+        AddonInfoList addonInfoList = reader.readFromXML(xml);
+        if (addonInfoList != null) {
+            candidateAddonInfos.addAll(addonInfoList.getAddons().stream()
+                    .filter(a -> !a.getDiscoveryMethods().isEmpty()).collect(Collectors.toSet()));
```

Comment:
I would not filter. Allowing all to pass through also provides the country and connection info to the AddonInfo, so we will be able to use it in the UI for filtering as well. That would solve the main issue preventing implementing https://github.com/openhab/openhab-webui/issues/2050

Reply:
Ok.

Reply:
Just thinking about this further. Currently we maintain a `Set<AddonInfo>` (both in the code in this PR and the OH-Core PR) whereby the unique filter for such set is the addon UID (the concatenation of addon type and name); however it is conceivable that some addons may be available via two addon providers (e.g. via OH main distribution and OH marketplace); and therefore two such `AddonInfo`s would be 'munged' into on entry in the set. => WDYT?

Note we do this 'munging' already in the case that an addon may have both a upnp and an mdns discovery method; whereby in this case it absolutely proper that there should be only one addon UID in the final suggestion set..

Reply:
See my comment here: https://github.com/openhab/openhab-core/pull/3806#issuecomment-1780795755

Reply:
The addon uid is different depending on the source of the addons. You are right for karaf provided addons. But for others, it may be something like: `marketplace:addonname`.

Reply:
Ok.

Reply:
I am not sure about the best solution yet. The registry return any addon with the id it finds in the providers: https://github.com/openhab/openhab-core/blob/f2c20af94458371adecd7adc4630312165fc9fed/bundles/org.openhab.core.addon/src/main/java/org/openhab/core/addon/AddonInfoRegistry.java#L71
That may have to be modified to return a "merged" version.
As described in the comment in the core PR, creating some kind of reference uid field so the suggestions can be linked to the original source could help with the merging.